### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Sites.Canonical

### PR DESCRIPTION
**Pull request by joelriou**
---

It compiles, but there is a `simp` issue discussed at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp.20regression/near/357814659

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Canonical.lean [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    intro Z W f h hf
+    apply (hB (U.downward_closed hf h)).isSeparatedFor.ext
+    intro Y l hl
+    apply (hB' hf (l ≫ h)).ext
+    intro M m hm
+    have : bind U B (m ≫ l ≫ h ≫ f) := by
+      -- porting note: had to make explicit the parameter `((m ≫ l ≫ h) ≫ f)` and
+      -- using `by exact`
+      have : bind U B ((m ≫ l ≫ h) ≫ f) := by exact Presieve.bind_comp f hf hm
+      -- porting note: was `simpa using this`, see also following porting notes
```
Comment (kmill) [inline Mathlib/CategoryTheory/Sites/Canonical.lean:93]: This seems to work for me:
```
have : bind U B ((m ≫ l ≫ h) ≫ f) := by exact Presieve.bind_comp f hf hm
simpa using this
```

File: Mathlib/CategoryTheory/Sites/Canonical.lean [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+      -- for similar problems with simp
+      convert this using 1
+      simp
+    -- porting note: without this strange `change _ = _`, the `trans` refused to work
+    change _ = _
```
Comment (kmill) [inline Mathlib/CategoryTheory/Sites/Canonical.lean:98]: This one is due to a known bug in `trans` that should be fixed in #3768

File: Mathlib/CategoryTheory/Sites/Canonical.lean (hunk 1 above)
Comment (kmill) [reply Mathlib/CategoryTheory/Sites/Canonical.lean:93]: By the way, for readability you can add the command
```
pp_extended_field_notation Sieve.arrows
```
to get `arrows` to show up with dot notation. It should go after the `structure` declaration.

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/Canonical.lean]: Thanks for these remarks!

**Comment by Ruben-VandeVelde** [comment]
@joelriou what are next steps here?

**Comment by joelriou** [comment]
@Ruben-VandeVelde As #3768 was merged, I could do a few fixes. This should be ready for review when CI finishes.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**